Use TCP keep alive to prevent connection freezing on slow connections V2

andreas-bok-sociomantic opened the pull request:
The mysql connection could indefinately freeze when in non-blocking mode
and on a bad internet connection. Set the TCP keep alive option to keep
the connection alive.

Extends the existing drizzle_options_st struct with members to needed to configure
the TCP keepalive for the socket connection.

andreas-bok-sociomantic committed:
Add socket options to drizzle_options_st

The added variables hold values needed to configure
the TCP keepalive for the socket connection.

andreas-bok-sociomantic committed:
Add methods to set and get socket options

andreas-bok-sociomantic committed:
Use TCP keep alive to prevent connection freezing on slow connections

The mysql connection could freeze indefinitely when in
non-blocking mode and on a bad internet connection.
Calls setsockopt with keepalive enabled to keep the
connection alive.

The options are set using the drizzle_socket_set_options or
drizzle_socket_set_option

andreas-bok-sociomantic committed:
Add unittest of methods to get and set socket options

Extends the existing unittest in connect.c with
testing of
- drizzle_socket_set_options
- drizzle_socket_set_option
- drizzle_socket_get_option

ben-palmer-sociomantic commented on libdrizzle-5.1/conn.h:
```diff
@@ -141,6 +141,66 @@ DRIZZLE_API
+ *
+ * @param[in] con Connection structure previously initialized with
+ *  drizzle_create(), drizzle_clone(), or related functions.
+ * @param[in] socket_option option to set the value for
```
Missing an `in` parameter in the documentation.

ben-palmer-sociomantic commented on libdrizzle/structs.h:
```diff
@@ -168,7 +172,11 @@ struct drizzle_options_st
-    socket_owner(DRIZZLE_SOCKET_OWNER_NATIVE)
+    socket_owner(DRIZZLE_SOCKET_OWNER_NATIVE),
+    wait_timeout(DRIZZLE_DEFAULT_SOCKET_TIMEOUT),
+    keepidle(5),
```
I wonder whether we should have `DRIZZLE_DEFAULT_x` values for the new parameters.

andreas-bok-sociomantic replied:
There are default values set by the system. Which we could use. Might be better in terms of having a "standard" linux configuration  when the values are not set explicitly
```
tcp_keepalive_time = 7200 (seconds)
tcp_keepalive_intvl = 75 (seconds)
tcp_keepalive_probes = 9 (number of probes)
```

ben-palmer-sociomantic commented on libdrizzle/conn.cc:
```diff
@@ -1621,11 +1622,67 @@ static drizzle_return_t _setsockopt(drizzle_st *con)
     return DRIZZLE_RETURN_ERRNO;
   }
 
+  ret = 1;
```
Do we need to set `ret` here?

ben-palmer-sociomantic replied:
or are we always setting keep alive?

andreas-bok-sociomantic replied:
Yes. Keep alive are enabled as default. The freeze  problem seemed to be the default values in linux that were used before this PR: 
 
tcp_keepalive_time = 7200 (seconds)
tcp_keepalive_intvl = 75 (seconds)
tcp_keepalive_probes = 9 (number of probes)

ben-palmer-sociomantic reviewed (commented):
Had an initial look, basically LGTM.

andreas-bok-sociomantic commented:
Changed initialization for socket options to linux default values 
```diff
diff --git a/libdrizzle/structs.h b/libdrizzle/structs.h
index 733144e..5d8d465 100644
--- a/libdrizzle/structs.h
+++ b/libdrizzle/structs.h
@@ -161,9 +161,9 @@ struct drizzle_options_st
   bool auth_plugin;
   drizzle_socket_owner socket_owner;
   int wait_timeout;
-  int keepidle;
-  int keepcnt;
-  int keepintvl;
+  int keepidle;  // default value under linux: 7200
+  int keepcnt;   // default value under linux: 75
+  int keepintvl; // default value under linux: 9
 
   drizzle_options_st() :
     non_blocking(false),
@@ -174,9 +174,9 @@ struct drizzle_options_st
     auth_plugin(false),
     socket_owner(DRIZZLE_SOCKET_OWNER_NATIVE),
     wait_timeout(DRIZZLE_DEFAULT_SOCKET_TIMEOUT),
-    keepidle(5),
-    keepcnt(3),
-    keepintvl(3)
+    keepidle(7200),
+    keepcnt(75),
+    keepintvl(9)
   { }
 };
 

```

andreas-bok-sociomantic commented:
Added missing `[in]` parameter

```diff
diff --git a/libdrizzle-5.1/conn.h b/libdrizzle-5.1/conn.h
index 10cc94f..2fb4c0c 100644
--- a/libdrizzle-5.1/conn.h
+++ b/libdrizzle-5.1/conn.h
@@ -172,8 +172,9 @@ void drizzle_socket_set_options(drizzle_options_st *options, int wait_timeout,
  *                                keepalive probes
  *
  * @param[in] con Connection structure previously initialized with
- *  drizzle_create(), drizzle_clone(), or related functions.
+ *                drizzle_create(), drizzle_clone(), or related functions.
  * @param[in] socket_option option to set the value for
+ * @param[in] value the value to set
  */
 DRIZZLE_API
 void drizzle_socket_set_option(drizzle_st *con, int socket_option, int value);

```

ben-palmer-sociomantic commented on libdrizzle/conn.cc:
```diff
@@ -220,6 +220,71 @@ void drizzle_options_destroy(drizzle_options_st *options)
+  switch (socket_option)
+  {
+    case DRIZZLE_SOCKET_TIMEOUT:
+      con->options.wait_timeout = value;
```
Perhaps we should have a `break;` at the end of each case statement.

ben-palmer-sociomantic commented:
One comment then LGTM.

andreas-bok-sociomantic commented:
- Fixed missing `break` stmts in switch. 

- Changed the socket options constants to a enum `drizzle_socket_option` to ensure set/get can only be called valid options parameter

Sorry I forgot to push the latter changes before the review

```diff 
diff --git a/libdrizzle-5.1/conn.h b/libdrizzle-5.1/conn.h
index 2fb4c0c..ff3ef1a 100644
--- a/libdrizzle-5.1/conn.h
+++ b/libdrizzle-5.1/conn.h
@@ -141,9 +141,9 @@ DRIZZLE_API
 void drizzle_options_destroy(drizzle_options_st *options);
 
 /**
- * Create a new socket options object with the specified values
+ * Sets several options for the socket connection
  *
- * @param[in] options The options object to set socket opts on
+ * @param[in] options An initialized options structure
  * @param[in] wait_timeout The timeout (in seconds) for setsockopt calls with
  *  option values: SO_SNDTIMEO, SO_RCVTIMEO, SO_LINGER
  * @param[in] keepidle The time (in seconds) the connection needs to remain
@@ -158,48 +158,42 @@ void drizzle_socket_set_options(drizzle_options_st *options, int wait_timeout,
                                 int keepidle, int keepcnt, int keepintvl);
 
 /**
- * Sets the value of the socket option. The available options to set are:
+ * Sets the value of a socket option. The available options to set are:
  *
- *  - DRIZZLE_SOCKET_TIMEOUT   : The timeout (in seconds) for setsockopt calls
- *                               with option values: SO_SNDTIMEO, SO_RCVTIMEO,
- *                               SO_LINGER
- *  - DRIZZLE_SOCKET_KEEPIDLE  : The time (in seconds) the connection needs to
- *                               remain idle before TCP starts sending keepalive
- *                               probes
- *  - DRIZZLE_SOCKET_KEEPCNT   : The maximum number of keepalive probes TCP
- *                               should send before dropping the connection.
- *  - DRIZZLE_SOCKET_KEEPINTVL : The time (in seconds) between individual
- *                                keepalive probes
+ *  DRIZZLE_SOCKET_OPTION_TIMEOUT : The timeout (in seconds) for setsockopt calls
+ *                                  with option values: SO_SNDTIMEO, SO_RCVTIMEO,
+ *                                  SO_LINGER
+ *
+ *  DRIZZLE_SOCKET_OPTION_KEEPIDLE : The time (in seconds) the connection needs
+ *                                   to remain idle before TCP starts sending
+ *                                   keepalive probes
+ *
+ *  DRIZZLE_SOCKET_OPTION_KEEPCNT : The maximum number of keepalive probes TCP
+ *                                  should send before dropping the connection.
+ *
+ *  DRIZZLE_SOCKET_OPTION_KEEPINTVL : The time (in seconds) between individual
+ *                                    keepalive probes
  *
  * @param[in] con Connection structure previously initialized with
  *                drizzle_create(), drizzle_clone(), or related functions.
- * @param[in] socket_option option to set the value for
+ * @param[in] option the option to set the value for
  * @param[in] value the value to set
  */
 DRIZZLE_API
-void drizzle_socket_set_option(drizzle_st *con, int socket_option, int value);
+void drizzle_socket_set_option(drizzle_st *con, drizzle_socket_option option,
+                               int value);
 
 /**
- * Gets the value of a socket option. The available options to get are:
- *
- *  - DRIZZLE_SOCKET_TIMEOUT   : The timeout (in seconds) for setsockopt calls
- *                               with option values: SO_SNDTIMEO, SO_RCVTIMEO,
- *                               SO_LINGER
- *  - DRIZZLE_SOCKET_KEEPIDLE  : The time (in seconds) the connection needs to
- *                               remain idle before TCP starts sending keepalive
- *                               probes
- *  - DRIZZLE_SOCKET_KEEPCNT   : The maximum number of keepalive probes TCP
- *                               should send before dropping the connection.
- *  - DRIZZLE_SOCKET_KEEPINTVL : The time (in seconds) between individual
- *                                keepalive probes
+ * Gets the value of a socket option. See drizzle_socket_set_options() for a
+ * description of the available options
  *
  * @param[in] con Connection structure previously initialized with
  *  drizzle_create(), drizzle_clone(), or related functions.
- * @param[in] socket_option option to get the value for
+ * @param[in] option option to get the value for
  * @return The value of the option, or -1 if the specified option doesn't exist
  */
 DRIZZLE_API
-int drizzle_socket_get_option(drizzle_st *con, int socket_option);
+int drizzle_socket_get_option(drizzle_st *con, drizzle_socket_option option);
 
 /**
  * Sets/unsets non-blocking connect option
diff --git a/libdrizzle-5.1/constants.h b/libdrizzle-5.1/constants.h
index a1cb12d..2a84f40 100644
--- a/libdrizzle-5.1/constants.h
+++ b/libdrizzle-5.1/constants.h
@@ -90,10 +90,6 @@ extern "C" {
 #define DRIZZLE_DEFAULT_SOCKET_TIMEOUT   10
 #define DRIZZLE_DEFAULT_SOCKET_SEND_SIZE DRIZZLE_DEFAULT_BUFFER_SIZE
 #define DRIZZLE_DEFAULT_SOCKET_RECV_SIZE DRIZZLE_DEFAULT_BUFFER_SIZE
-#define DRIZZLE_SOCKET_KEEPIDLE          4
-#define DRIZZLE_SOCKET_KEEPCNT           5
-#define DRIZZLE_SOCKET_KEEPINTVL         6
-#define DRIZZLE_SOCKET_TIMEOUT           7
 #define DRIZZLE_MYSQL_PASSWORD_HASH      41
 #define DRIZZLE_BINLOG_CRC32_LEN         4
 // If this version or higher then we are doing checksums
@@ -295,6 +291,18 @@ typedef enum
   DRIZZLE_SOCKET_OWNER_CLIENT
 } drizzle_socket_owner;
 
+/**
+ * @ingroup drizzle_con
+ * Available options to set for the socket connection
+ */
+typedef enum
+{
+  DRIZZLE_SOCKET_OPTION_KEEPIDLE=4,
+  DRIZZLE_SOCKET_OPTION_KEEPCNT,
+  DRIZZLE_SOCKET_OPTION_KEEPINTVL,
+  DRIZZLE_SOCKET_OPTION_TIMEOUT
+} drizzle_socket_option;
+
 typedef enum
 {
   DRIZZLE_EVENT_POSITION_TIMESTAMP= 0,
diff --git a/libdrizzle/conn.cc b/libdrizzle/conn.cc
index ed57dd6..c18b278 100644
--- a/libdrizzle/conn.cc
+++ b/libdrizzle/conn.cc
@@ -233,51 +233,56 @@ void drizzle_socket_set_options(drizzle_options_st *options, int wait_timeout,
   options->keepintvl = keepintvl;
 }
 
-void drizzle_socket_set_option(drizzle_st *con, int socket_option, int value)
+void drizzle_socket_set_option(drizzle_st *con, drizzle_socket_option option,
+  int value)
 {
   if (con == NULL)
   {
     return;
   }
 
-  switch (socket_option)
+  switch (option)
   {
-    case DRIZZLE_SOCKET_TIMEOUT:
+    case DRIZZLE_SOCKET_OPTION_TIMEOUT:
       con->options.wait_timeout = value;
+      break;
 
-    case DRIZZLE_SOCKET_KEEPIDLE:
+    case DRIZZLE_SOCKET_OPTION_KEEPIDLE:
       con->options.keepidle = value;
+      break;
 
-    case DRIZZLE_SOCKET_KEEPCNT:
+    case DRIZZLE_SOCKET_OPTION_KEEPCNT:
       con->options.keepcnt = value;
+      break;
 
-    case DRIZZLE_SOCKET_KEEPINTVL:
+    case DRIZZLE_SOCKET_OPTION_KEEPINTVL:
       con->options.keepintvl = value;
+      break;
 
     default:
       break;
   }
 }
 
-int drizzle_socket_get_option(drizzle_st *con, int socket_option)
+int drizzle_socket_get_option(drizzle_st *con, drizzle_socket_option option)
 {
   if (con == NULL)
   {
     return -1;
   }
 
-  switch (socket_option)
+  switch (option)
   {
-    case DRIZZLE_SOCKET_TIMEOUT:
+    case DRIZZLE_SOCKET_OPTION_TIMEOUT:
       return con->options.wait_timeout;
 
-    case DRIZZLE_SOCKET_KEEPIDLE:
+    case DRIZZLE_SOCKET_OPTION_KEEPIDLE:
       return con->options.keepidle;
 
-    case DRIZZLE_SOCKET_KEEPCNT:
+    case DRIZZLE_SOCKET_OPTION_KEEPCNT:
       return con->options.keepcnt;
 
-    case DRIZZLE_SOCKET_KEEPINTVL:
+    case DRIZZLE_SOCKET_OPTION_KEEPINTVL:
       return con->options.keepintvl;
 
     default:
diff --git a/tests/unit/connect.c b/tests/unit/connect.c
index a7e37db..eca9f74 100644
--- a/tests/unit/connect.c
+++ b/tests/unit/connect.c
@@ -57,24 +57,24 @@ int main(int argc, char *argv[])
                                   getenv("MYSQL_SCHEMA"), opts);
   ASSERT_NOT_NULL_(con, "Drizzle connection object creation error");
 
-  int opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_TIMEOUT);
+  int opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_OPTION_TIMEOUT);
   ASSERT_EQ_(10, opt_val, "unexpected value for socket option TIMEOUT: %d != 10",
     opt_val);
 
-  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_KEEPIDLE);
+  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_OPTION_KEEPIDLE);
   ASSERT_EQ_(5, opt_val, "unexpected value for socket option KEEPIDLE: %d != 5",
     opt_val);
 
-  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_KEEPCNT);
+  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_OPTION_KEEPCNT);
   ASSERT_EQ_(3, opt_val, "unexpected value for socket option KEEPCNT: %d != 3",
     opt_val);
 
-  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_KEEPINTVL);
+  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_OPTION_KEEPINTVL);
   ASSERT_EQ_(3, opt_val, "unexpected value for socket option KEEPINTVL: %d != 3",
     opt_val);
 
-  drizzle_socket_set_option(con, DRIZZLE_SOCKET_TIMEOUT, 20);
-  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_TIMEOUT);
+  drizzle_socket_set_option(con, DRIZZLE_SOCKET_OPTION_TIMEOUT, 20);
+  opt_val = drizzle_socket_get_option(con, DRIZZLE_SOCKET_OPTION_TIMEOUT);
   ASSERT_EQ_(20, opt_val, "unexpected value for socket option KEEPALIVE: %d != 20",
     opt_val);
```

andreas-bok-sociomantic commented:
Rebased on latest v5.x.x branch

ben-palmer-sociomantic commented:
Cool, LGTM.
